### PR TITLE
[link-raw] fix radio with multiple transmit buffers

### PR DESCRIPTION
**Pull request by bukepo**
`otPlatRadioGetTransmitBuffer()` may return different values and even null on certain platform. This PR changed link raw so that this API is only called once. Otherwise, it may result in null pointer crash.
Besides,
* avoided reentrance issue of InvokeTransmitDone by deferring transmit in a tasklet.
* corrected how software retransmissions behavior when radio itself supports retransmission.

### CODE REVIEW

File: src/core/mac/link_raw.cpp [hunk 1]
```diff
@@ -70,7 +72,22 @@ LinkRaw::LinkRaw(Instance &aInstance)
     , mEnergyScanDoneCallback(NULL)
 {
     // Query the capabilities to check asserts
-    (void)GetCaps();
+    mRadioCaps     = otPlatRadioGetCaps(&GetInstance());
```
Comment (abtink) [inline src/core/mac/link_raw.cpp]: Suggestion: How about adding `mRadioCaps` and `mTransmitFrame` to the constructor initializer list?

File: src/core/mac/link_raw.cpp [hunk 2]
```diff
@@ -224,117 +238,111 @@ void LinkRaw::InvokeReceiveDone(otRadioFrame *aFrame, otError aError)
...
 
 #if OPENTHREAD_CONFIG_ENABLE_SOFTWARE_ACK_TIMEOUT
-    mTimer.Stop();
+    if ((mRadioCaps & OT_RADIO_CAPS_ACK_TIMEOUT) == 0)
+    {
+        mTimer.Stop();
+    }
 #endif // OPENTHREAD_CONFIG_ENABLE_SOFTWARE_ACK_TIMEOUT
 
 #if OPENTHREAD_CONFIG_ENABLE_SOFTWARE_RETRANSMIT
```
Comment (abtink) [inline src/core/mac/link_raw.cpp]: What happens if radio provides support for CSMA backoff but not retries and we have `SOFTWARE_RETRANSMIT` enabled?
In that case, if the error `CHANNEL_ACCESS_FAILURE` we should not do `mCsmaBackOff++` and the CSMA retry and directly go to check for retx. 

I suggest keeping the logic separate, i.e., we first check if sw CSMA backoff is needed (`SOFTWARE_CSMA_BACKOFF` is enabled and radio does no provide it) and handle CSMA retry and then check if retx is needed and handle it (sort of like the `mac.cpp` implementation [here](https://github.com/openthread/openthread/blob/master/src/core/mac/mac.cpp#L1346))

**Comment by bukepo** [comment]
@abtink I have modified according to your comments. Thank you.

File: src/core/mac/link_raw.cpp [hunk 3]
```diff
@@ -224,117 +241,130 @@ void LinkRaw::InvokeReceiveDone(otRadioFrame *aFrame, otError aError)
...
+    }
+}
+
+void LinkRaw::TransmitNow(void)
+{
+    otError error = otPlatRadioTransmit(&GetInstance(), mTransmitFrame);
+
+    if (error != OT_ERROR_NONE)
+    {
+        InvokeTransmitDone(mTransmitFrame, NULL, error);
```
Comment (abtink) [inline src/core/mac/link_raw.cpp]: How about using `OT_ERROR_ABORT` in TransmitDone?

Comment (bukepo) [reply src/core/mac/link_raw.cpp]: I will update the code.

File: src/core/mac/link_raw.hpp [hunk 4]
```diff
@@ -237,13 +242,31 @@ class LinkRaw
...
-    otInstance &mInstance;
+    enum
+    {
+        kOperationTransmitData = 1 << 0,
+    };
+
+    void        TransmitNow(void);
+    void        StartTransmit(void);
+    static void HandleOperationTask(Tasklet &aTasklet);
+    void        HandleOperationTask();
```
Comment (jwhui) [inline src/core/mac/link_raw.hpp]: All void functions or methods shall explicitly declare and specify the void type keyword.

File: src/core/mac/link_raw.hpp [hunk 5]
```diff
@@ -237,13 +242,31 @@ class LinkRaw
...
+
+    Tasklet  mOperationTask;
+    uint16_t mPendingOperations;
 
 #if OPENTHREAD_LINKRAW_TIMER_REQUIRED
 
-    enum TimerReason
-    {
+    enum TimerReason{
```
Comment (jwhui) [inline src/core/mac/link_raw.hpp]: Is this change intentional?

File: src/core/mac/link_raw.hpp [hunk 6]
```diff
@@ -237,13 +242,31 @@ class LinkRaw
...
+        kOperationTransmitData = 1 << 0,
+    };
+
+    void        TransmitNow(void);
+    void        StartTransmit(void);
+    static void HandleOperationTask(Tasklet &aTasklet);
+    void        HandleOperationTask();
+
+    Tasklet  mOperationTask;
+    uint16_t mPendingOperations;
```
Comment (jwhui) [inline src/core/mac/link_raw.hpp]: Are there plans to add more pending operations? If not, declare this as type bool and 1-bit bit field?

File: src/core/mac/link_raw.cpp [hunk 7]
```diff
@@ -159,40 +179,37 @@ otError LinkRaw::SetShortAddress(uint16_t aShortAddress)
...
 {
-    otRadioCaps RadioCaps = otPlatRadioGetCaps(&mInstance);
-
-    // The radio shouldn't support a capability if it is being compile
-    // time included into the raw link-layer code.
+    otRadioCaps radioCaps = mRadioCaps;
 
 #if OPENTHREAD_CONFIG_ENABLE_SOFTWARE_ACK_TIMEOUT
-    if ((RadioCaps & OT_RADIO_CAPS_ACK_TIMEOUT) == 0)
+    if ((mRadioCaps & OT_RADIO_CAPS_ACK_TIMEOUT) == 0)
```
Comment (jwhui) [inline src/core/mac/link_raw.cpp]: What is the purpose of these conditionals? Why not just set the bit whenever the build option is included?

File: src/core/mac/link_raw.cpp [hunk 8]
```diff
@@ -224,117 +241,142 @@ void LinkRaw::InvokeReceiveDone(otRadioFrame *aFrame, otError aError)
...
+    }
+}
+
 otError LinkRaw::Transmit(otRadioFrame *aFrame, otLinkRawTransmitDone aCallback)
 {
-    otError error = OT_ERROR_INVALID_STATE;
+    otError error = OT_ERROR_NONE;
 
-    if (mEnabled)
+    assert(mTransmitFrame == aFrame);
```
Comment (jwhui) [inline src/core/mac/link_raw.cpp]: If `aFrame` must be `mTransmitFrame`, should we simply remove the `aFrame` parameter?

File: src/core/mac/link_raw.hpp (hunk 5 above)
Comment (bukepo) [reply src/core/mac/link_raw.hpp]: No, it's caused by my local clang-format-6.0. I have moved the macro location and it magically get right now.

File: src/core/mac/link_raw.hpp (hunk 6 above)
Comment (bukepo) [reply src/core/mac/link_raw.hpp]: I thought this class could be finally be base class of Mac. Declaring this way could share this variable with its subclass. I have modified according to your suggestion.
